### PR TITLE
add name and description to SSM Maintenance Window Tasks

### LINKED ISSUE #5702: Add name and description to SSM Maintenance Window Tasks
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Add the name and description fields in terraform when creating SSM Maintenance window tasks

The console allows to add a name and description to tasks. Terraform doesn't support these 2 fields.

* aws_ssm_maintenance_window_task

### Potential Terraform Configuration

```hcl
resource "aws_ssm_maintenance_window_task" "maintenance_task" {
  name                  = ""
  description           = ""
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by Guimove**
Fixes #5702

Changes proposed in this pull request:

* Add the name and description fields in terraform when creating SSM Maintenance window tasks

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSSSMMaintenanceWindowTarget_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSSSMMaintenanceWindowTarget_basic -timeout 120m
=== RUN   TestAccAWSSSMMaintenanceWindowTarget_basic
--- PASS: TestAccAWSSSMMaintenanceWindowTarget_basic (20.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	20.497s

...
```
@Ninir : Please be kind, this is my first contrib :)

### CODE REVIEW

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 1]
```diff
@@ -73,6 +73,18 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
```
Comment (Ninir) [inline aws/resource_aws_ssm_maintenance_window_task.go]: As per [the related documentation](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_RegisterTaskWithMaintenanceWindow.html#systemsmanager-RegisterTaskWithMaintenanceWindow-request-Description), this is missing the Validation Constraints which are:

> Minimum length of 1. Maximum length of 128.

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 2]
```diff
@@ -73,6 +73,18 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 				},
 			},
 
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
```
Comment (Ninir) [inline aws/resource_aws_ssm_maintenance_window_task.go]: As per [the related documentation](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_RegisterTaskWithMaintenanceWindow.html#systemsmanager-RegisterTaskWithMaintenanceWindow-request-Name), this is missing the Validation Constraints which are:

> Length Constraints: Minimum length of 3. Maximum length of 128.
> Pattern: ^[a-zA-Z0-9_\-.]{3,128}$

File: aws/resource_aws_ssm_maintenance_window_task_test.go [hunk 3]
```diff
@@ -141,6 +141,8 @@ resource "aws_ssm_maintenance_window_task" "target" {
   task_type = "RUN_COMMAND"
   task_arn = "AWS-RunShellScript"
   priority = 1
+  name = "TestMaintenanceWindowTask"
```
Comment (Ninir) [inline aws/resource_aws_ssm_maintenance_window_task_test.go]: In order to validate that these are actually set and retrieved, we should extend [the related acceptance test](https://github.com/terraform-providers/terraform-provider-aws/blob/2728e9d7828fd75ff07a1d539f8295f7c3fdcf58/aws/resource_aws_ssm_maintenance_window_task_test.go#L27) with something like:

```go
func TestAccAWSSSMMaintenanceWindowTask_basic(t *testing.T) {
	var task ssm.MaintenanceWindowTask

	name := acctest.RandString(10)
	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSSSMMaintenanceWindowTaskDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAWSSSMMaintenanceWindowTaskBasicConfig(name),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSSSMMaintenanceWindowTaskExists("aws_ssm_maintenance_window_task.target", &task),
resource.TestCheckResourceAttr("aws_ssm_maintenance_window_task.target", "name", "TestMaintenanceWindowTask"),
resource.TestCheckResourceAttr("aws_ssm_maintenance_window_task.target", "description", "This ressource is for test purpose only"),
				),
			},
		},
	})
}
```

File: aws/resource_aws_ssm_maintenance_window_task.go (hunk 1 above)
Comment (Guimove) [reply aws/resource_aws_ssm_maintenance_window_task.go]: Created the validator for this

File: aws/resource_aws_ssm_maintenance_window_task.go (hunk 2 above)
Comment (Guimove) [reply aws/resource_aws_ssm_maintenance_window_task.go]: Created the validator for this

File: aws/resource_aws_ssm_maintenance_window_task_test.go (hunk 3 above)
Comment (Guimove) [reply aws/resource_aws_ssm_maintenance_window_task_test.go]: Seems working with your example @Ninir 
thanks

**Comment by Guimove** [comment]
Hello @Ninir, as discussed with you, I've added the two new validators and updated the test to check the name and description.
Thanks

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 4]
```diff
@@ -73,6 +73,20 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAwsSSMMaintenanceWindowTaskName,
+			},
+
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAwsSSMMaintenanceWindowTaskDescription,
```
[the remark this answers is not in the document]
Comment (Guimove) [reply aws/resource_aws_ssm_maintenance_window_task.go]: Ok for the description but for the name I can keep my validator that check the regex and the length ?

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 5]
```diff
@@ -241,6 +257,14 @@ func resourceAwsSsmMaintenanceWindowTaskRead(d *schema.ResourceData, meta interf
 			d.Set("task_arn", t.TaskArn)
 			d.Set("priority", t.Priority)
 
+			if t.Name != nil {
```
[the remark this answers is not in the document]
Comment (Guimove) [reply aws/resource_aws_ssm_maintenance_window_task.go]: Ok I'll remove that if

File: aws/resource_aws_ssm_maintenance_window_task_test.go [hunk 6]
```diff
@@ -141,6 +143,8 @@ resource "aws_ssm_maintenance_window_task" "target" {
...
+  description = "This ressource is for test purpose only"
```
[the remark this answers is not in the document]
Comment (Guimove) [reply aws/resource_aws_ssm_maintenance_window_task_test.go]: This is the french touch hehe

**Comment by Guimove** [comment]
Updates are done :crossed_fingers:

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 7]
```diff
@@ -73,6 +74,20 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 				},
 			},
 
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
```
Comment (Ninir) [inline aws/resource_aws_ssm_maintenance_window_task.go]: Is this really forcing a new resource? The API seems to tell the opposite (i.e. it's updatable): https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_UpdateMaintenanceWindow.html

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 8]
```diff
@@ -73,6 +74,20 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
+				ValidateFunc: validateAwsSSMMaintenanceWindowTaskName,
+			},
+
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
```
Comment (Ninir) [inline aws/resource_aws_ssm_maintenance_window_task.go]: Same here

**Comment by bflad** [comment]
This has been released in [version 1.44.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1440-november-14-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by YakDriver** [comment]
Please see #7186 which fixes this PR.

**Pull request merged**